programs/vote/src/vote_state.rs

```diff
@@ -386,6 +414,127 @@ impl VoteState {
+            let epoch_of_last_authorized_switch =
+                self.prior_voters.last().map(|range| range.2).unwrap_or(0);
+
+            // leader_schedule_epoch must:
+            // 1) Be monotonically increasing due to the clock always
+            //    moving forward
+            // 2) not be equal to latest epoch otherwise this
+            //    function would have returned TooSoonToReauthorize error
+            //    above
+            assert!(leader_schedule_epoch > *latest_epoch);
+
+            // Commit the new state
+            self.prior_voters.append((
+                *latest_authorized_pubkey,
+                epoch_of_last_authorized_switch,
+                leader_schedule_epoch,
+            ));
+        }
+
+        self.authorized_voter
+            .insert(leader_schedule_epoch, *authorized_pubkey);
+
+        Ok(())
+    }
+
+    fn get_and_update_authorized_voter(&mut self, current_epoch: u64) -> Option<Pubkey> {
+        let (pubkey, existed) = self
+            .get_or_calculate_authorized_voter_for_epoch(current_epoch)
+            .expect(
+                "Internal functions should
+        only call this will monotonically increasing current_epoch",
+            );
+        if !existed {
+            self.authorized_voter.insert(current_epoch, pubkey);
+        }
+        self.update_authorized_voters(current_epoch);
+        Some(pubkey)
+    }
+
+    // Returns the authorized voter at the given epoch if the epoch is >= the
+    // current epoch, and a bool indicating whether the entry for this epoch
+    // exists in the self.authorized_voter map
+    fn get_or_calculate_authorized_voter_for_epoch(&self, epoch: u64) -> Option<(Pubkey, bool)> {
+        let res = self.authorized_voter.get(&epoch);
+        if res.is_none() {
+            // If no authorized voter has been set yet for this epoch,
+            // this must mean the authorized voter remains unchanged
+            // from the latest epoch before this one
+            let res = self.authorized_voter.range(0..epoch).next_back();
+
+            if res.is_none() {
+                warn!(
+                    "Tried to query for the authorized voter of an epoch earlier
+                    than the current epoch. Earlier epochs have been purged"
+                );
+            }
+
+            res.map(|(_, pubkey)| (*pubkey, false))
+        } else {
+            res.map(|pubkey| (*pubkey, true))
+        }
+    }
+
+    fn update_authorized_voters(&mut self, current_epoch: u64) -> bool {
+        // Iterate through the keys in order, filtering out the ones
+        // less than the current epoch
+        let expired_keys: Vec<_> = self
+            .authorized_voter
+            .range(0..current_epoch)
+            .map(|(authorized_epoch, _)| *authorized_epoch)
+            .collect();
+
+        for key in expired_keys {
+            self.authorized_voter.remove(&key);
+        }
+
+        // Have to uphold this invariant b/c this is
```

Comment:
```suggestion
        // Have to uphold this invariant because this is
```
(don't abbreviate, use full words)